### DISCUSSION TITLE
The new grid list view idea

### DISCUSSION THREAD

**Discussion by heyarny** [shyndman/flutter_layout_grid#65] [Ideas]
Discussion from https://github.com/madewithfelt/flutter_layout_grid/issues/64

Let's see :)

**Comment by heyarny** [comment]
There are actually many blogs about it already.
Check this one for instance: https://blog.logrocket.com/virtual-scrolling-core-principles-and-basic-implementation-in-react/

**Reply by shyndman** [reply]
Great article. These are actually the same principles that underly the sliver architecture in Flutter (well...some of it...it actually deals with all sorts of interesting scroll problems).

If you haven't, you should run the example provided with the `scrollable_positioned_list` library. It looks like we can get first and last displayed index out of it, AND it handles jumping past items of unknown height (bit of a bug there, jumping to the end of the list, but no biggie).


https://user-images.githubusercontent.com/42326/130706450-cef0dbf5-d3c7-4852-9aeb-813cf0998812.mov

**Reply by shyndman** [reply]
I also managed to add a `Scrollbar`, which acts a little oddly, but I'm confident the experience could be improved somewhat.

**Reply by heyarny** [reply]
That scroll animation looks kinda weird.
There are 5000 items in the list which are not all attached and scroll-to works just fine?
How do they know the distance to scroll?

I looked up the source code of the list and there are still a lot of things I don't understand.

**Reply by heyarny** [reply]
Ok, I can confirm it seems to work, except scrollbar is behaving strange.
There's already an issue for that https://github.com/google/flutter.widgets/issues/273

**Reply by heyarny** [reply]
This guy is having the same problem: https://github.com/google/flutter.widgets/issues/145 (1year+ old)
Seems like flutter is not there yet to have everything in one widget.
I guess ScrollablePositionedList is not their priority as its latest release is still not null-safety.

**Reply by shyndman** [reply]
> That scroll animation looks kinda weird.

It's running in debug mode, which is considerably slower. I've attached a version built in release mode (at the bottom). Does it look better to you?

> There are 5000 items in the list which are not all attached and scroll-to works just fine?
> How do they know the distance to scroll?

Yup.

Looks like they're winging it, and correcting after the fact. I added a `ScrollBar` so I could see how it behaves, and its clear that if you jump past a bunch of rows then scroll up, the height of the viewport and scroll position is being recomputed regularly (ie, the scrollbar jumps around, even if you're scrolling up).

I've had an opportunity to read over the code quickly, but I need to sit down with it to get a better idea.

https://user-images.githubusercontent.com/42326/130809920-7d39907d-47ec-4f9c-9efd-37d136120ee1.mov

**Reply by heyarny** [reply]
I meant the "fading"-part while scrolling is weird. I guess they're pre-creating the block of rows you're about to scroll to and simply fade into that block. So actually smarter instead of scrolling through all the rows and create unnecessary widgets.

I'll try to find other ways to get to the goal. Maybe with slivers or something else.

**Comment by heyarny** [comment]
@shyndman look what I've found: https://gist.github.com/aloisdeniel/81d6a5241ca9e2f82f17cf3c001444c5

This sliver grid layout is working great so far as a base layout.
I like it as this seems to be the most state-of-the-art way to create layouts using grid-view and slivers.

The layouting is working fine so far and now I'm trying to solve the following:
- auto extent height to adjust height to widgets content (I have an idea, but not sure yet)
- insert/deletion animations (might be easier, as this solution is based on gridview/sliver standards)
- scroll to/jump to (might be possible, if we know all the positions & sizes of all widgets)

**Reply by shahmirzali49** [reply]
@heyarny finally did you find a solution or ? :)

**Reply by heyarny** [reply]
@Shahmirzali-Huseynov nope, I gave up on flutter. Good luck

**Reply by ghost** [reply]
> @Shahmirzali-Huseynov nope, I gave up on flutter. Good luck

Sad (, are you using other cross platform or native?

**Reply by heyarny** [reply]
Take a look at KMM https://kotlinlang.org/docs/multiplatform-mobile-getting-started.html
I'm still on it, but is seems very promising.
